ortools/sat/csharp/CpModel.cs

```diff
@@ -99,37 +95,51 @@ public Constraint AddLinearExpressionInDomain(LinearExpr linear_expr, Domain dom
             linear.Vars.Add(term.Key.Index);
             linear.Coeffs.Add(term.Value);
         }
-        foreach (long value in domain.FlattenedIntervals())
+
+        linear.Domain.Add(lb is Int64.MinValue or Int64.MaxValue ? lb : lb - constant);
+        linear.Domain.Add(ub is Int64.MinValue or Int64.MaxValue ? ub : ub - constant);
+
+        ct.Proto.Linear = linear;
+        return ct;
+    }
+
+    public Constraint AddLinearExpressionInDomain(LinearExpr linear_expr, long[] flattenedIntervals)
+    {
+        Dictionary<IntVar, long> dict = new Dictionary<IntVar, long>();
+        long constant = LinearExpr.GetVarValueMap(linear_expr, 1L, dict);
+        Constraint ct = new Constraint(model_);
+        LinearConstraintProto linear = new LinearConstraintProto();
+        foreach (KeyValuePair<IntVar, long> term in dict)
         {
-            if (value == Int64.MinValue || value == Int64.MaxValue)
-            {
-                linear.Domain.Add(value);
-            }
-            else
-            {
-                linear.Domain.Add(value - constant);
-            }
+            linear.Vars.Add(term.Key.Index);
+            linear.Coeffs.Add(term.Value);
+        }
+        foreach (long value in flattenedIntervals)
+        {
+            linear.Domain.Add(value is Int64.MinValue or Int64.MaxValue ? value : value - constant);
         }
         ct.Proto.Linear = linear;
         return ct;
     }
 
+    public Constraint AddLinearExpressionInDomain(LinearExpr linear_expr, Domain domain)
+        => AddLinearExpressionInDomain(linear_expr, domain.FlattenedIntervals());
+
     public Constraint Add(BoundedLinearExpression lin)
     {
         switch (lin.CtType)
         {
         case BoundedLinearExpression.Type.BoundExpression: {
-            return AddLinearExpressionInDomain(lin.Left, new Domain(lin.Lb, lin.Ub));
+            return AddLinearExpressionInDomain(lin.Left, lin.Lb, lin.Ub);
```

Comment:
use AddLinearExpression(expr, lb, ub) and optimize the code there.

Reply:
Is there such a method? 🤔 Or did you mean to rename them?

Reply:
It is called AddLinearConstraint(). We can optimize the code in there not to create a domain.